🐛 Number of crossings calculation

simon1hofmann opened the pull request:
## Description

Deleting a wire can decrease the number of crossings in two ways:
1. Deleted wire was in the crossing layer and had a wire in the ground layer
2. Deleted wire was in the ground layer and had a wire in the crossing layer

Similar adding a wire can increase the number of crossings in two ways:
1. Added wire is in the crossing layer and there is already a wire in the ground layer
2. Added wire is in the ground layer and there is already a wire in the crossing layer

The previous implementation only covered the first cases.
Thanks a lot to @Drewniok and @samuelshng for spotting this bug.

## Checklist:
- [x] The pull request only contains commits that are related to it.
- [x] I have added appropriate tests and documentation.
- [ ] I have added a changelog entry.
- [ ] I have created/adjusted the Python bindings for any new or updated functionality.
- [ ] I have made sure that all CI jobs on GitHub pass.
- [x] The pull request introduces no new warnings and follows the project's style guidelines.

marcelwa commented:
> ## Description
> Deleting a wire can decrease the number of crossings in two ways:
> 
> 1. Deleted wire was in the crossing layer and had a wire in the ground layer
> 2. Deleted wire was in the ground layer and had a wire in the crossing layer
> 
> Similar adding a wire can increase the number of crossings in two ways:
> 
> 1. Added wire is in the crossing layer and there is already a wire in the ground layer
> 2. Added wire is in the ground layer and there is already a wire in the crossing layer

@simon1hofmann my understanding was that when a wire gets deleted in the ground layer, with another wire remaining in the crossing layer, the crossing layer wire should be moved to the ground layer. What do you think?

simon1hofmann commented:
> > ## Description
> > Deleting a wire can decrease the number of crossings in two ways:
> > 
> > 1. Deleted wire was in the crossing layer and had a wire in the ground layer
> > 2. Deleted wire was in the ground layer and had a wire in the crossing layer
> > 
> > Similar adding a wire can increase the number of crossings in two ways:
> > 
> > 1. Added wire is in the crossing layer and there is already a wire in the ground layer
> > 2. Added wire is in the ground layer and there is already a wire in the crossing layer
> 
> @simon1hofmann my understanding was that when a wire gets deleted in the ground layer, with another wire remaining in the crossing layer, the crossing layer wire should be moved to the ground layer. What do you think?

Yes, that should be the default in my opinion. Currently, `post_layout_optimization` does that after deleting wires in the ground layer with another wire in the crossing layer:
```
    void fix_wires(ObstrLyt& lyt, const std::vector<tile<ObstrLyt>>& deleted_coords) noexcept
    {
        for (const auto& tile : deleted_coords)
        {
            const auto ground = lyt.below(tile);
            const auto above  = lyt.above(tile);

            if (lyt.is_empty_tile(ground) && lyt.is_wire_tile(above))
            {
                const auto incoming_tile = lyt.incoming_data_flow(above).front();
                const auto outgoing_tile = lyt.outgoing_data_flow(above).front();

                // move wire from z=1 to z=0
                lyt.move_node(lyt.get_node(above), ground, {lyt.make_signal(lyt.get_node(incoming_tile))});

                // if outgoing tile has other incoming signals (e.g. AND), update children
                if (const auto in_flow = lyt.incoming_data_flow(outgoing_tile); !in_flow.empty())
                {
                    const auto front = in_flow.front();

                    if (std::find(deleted_coords.cbegin(), deleted_coords.cend(), front) == deleted_coords.cend() ||
                        std::find(moved_tiles.cbegin(), moved_tiles.cend(), front) != moved_tiles.cend())
                    {
                        lyt.move_node(lyt.get_node(outgoing_tile), outgoing_tile,
                                      {lyt.make_signal(lyt.get_node(ground)), lyt.make_signal(lyt.get_node(front))});
                    }
                }
                // otherwise, the wire is the only incoming signal
                else
                {
                    lyt.move_node(lyt.get_node(outgoing_tile), outgoing_tile, {lyt.make_signal(lyt.get_node(ground))});
                }
            }
        }
    }
    ```

marcelwa reviewed (approved):
Many thanks for the quick fix! :pray:
